### PR TITLE
fix: pull should return 404 when image not exists

### PR DESCRIPTION
**Pull request by ZYecho**
Signed-off-by: zhangyue <zy675793960@yeah.net>

### Ⅰ. Describe what this PR did
fix bugs in #2913, when error happens, we should not write anything to Writer, if you write anything, then we will get resp code 200.
In this pr, add a `resolveImage` interface to check the ref is exists or not, if not, will return err.

### Ⅱ. Does this pull request fix one issue?

fix bugs in #2913

### Ⅲ. Why don't you add test cases (unit test/integration test)? (你真的觉得不需要加测试吗？)
takes on #2913 

### Ⅳ. Describe how to verify it
wait for CI

### Ⅴ. Special notes for reviews
when image pull No.5 layer error,  the error msg still in json stream, this pr doesn't handle this case.

### CODE REVIEW

**Comment by allencloud** [comment]
PTAL @fuweid thanks a lot.

File: apis/server/image_bridge.go [hunk 1]
```diff
@@ -58,7 +58,7 @@ func (s *Server) pullImage(ctx context.Context, rw http.ResponseWriter, req *htt
 	// Error information has be sent to client, so no need call resp.Write
 	if err := s.ImageMgr.PullImage(ctx, image, &authConfig, newWriteFlusher(rw)); err != nil {
 		logrus.Errorf("failed to pull image %s: %v", image, err)
-		return nil
+		return err
```
Comment (fuweid) [inline apis/server/image_bridge.go]: well. if we write something into response, the http code will be 200 which can never be changed.
so we use jsonstream to return error. the client needs to parse the json stream to get error.

File: test/api_image_create_test.go [hunk 2]
```diff
@@ -47,6 +47,19 @@ func (suite *APIImageCreateSuite) TestImageCreateNil(c *check.C) {
...
+func (suite *APIImageCreateSuite) TestImageCreateNonExistentImage(c *check.C) {
+	q := url.Values{}
+	image := "qwefghjm:zxcvbn_efgh_nonexist"
+
+	q.Add("fromImage", image)
+	query := request.WithQuery(q)
+
+	resp, err := request.Post("/images/create", query)
+	c.Assert(err, check.IsNil)
+	CheckRespStatus(c, resp, 500)
```
Comment (fuweid) [inline test/api_image_create_test.go]: we need to parse the json stream here

File: apis/server/image_bridge.go (hunk 1 above)
Comment (fuweid) [reply apis/server/image_bridge.go]: when you pull the No. 5 layer and get something wrong like connection refuse, your change is not helpful. it always be 200, not 500

**Comment by ZYecho** [comment]
@fuweid PTAL

File: test/api_image_create_test.go (hunk 2 above)
Comment (fuweid) [inline test/api_image_create_test.go]: is it 400? not 500?

File: ctrd/image.go [hunk 3]
```diff
@@ -251,21 +251,24 @@ func (c *Client) PushImage(ctx context.Context, ref string, authConfig *types.Au
...
 
-// FetchImage fetches image content from the remote repository.
-func (c *Client) FetchImage(ctx context.Context, name string, refs []string, authConfig *types.AuthConfig, stream *jsonstream.JSONStream) (containerd.Image, error) {
-	wrapperCli, err := c.Get(ctx)
+// ResolveImage attempts to resolve the image reference into a available reference and resolver.
+func (c *Client) ResolveImage(ctx context.Context, nameRef string, refs []string, authConfig *types.AuthConfig, opts docker.ResolverOptions) (remotes.Resolver, string, error) {
+	resolver, availableRef, err := c.getResolver(ctx, authConfig, nameRef, refs, opts)
 	if err != nil {
-		return nil, fmt.Errorf("failed to get a containerd grpc client: %v", err)
+		return nil, "", err
```
Comment (fuweid) [inline ctrd/image.go]: I think we should check the status and make sure that it is 404

Comment (ZYecho) [reply ctrd/image.go]: when resolveImage return err, means that it can't find available ref to use.

File: cri/v1alpha2/cri.go [hunk 4]
```diff
@@ -1359,6 +1361,9 @@ func (c *CriManager) PullImage(ctx context.Context, r *runtime.PullImageRequest)
 	}
 
 	if err := c.ImageMgr.PullImage(ctx, imageRef, authConfig, bytes.NewBuffer([]byte{})); err != nil {
+		if err == errtypes.ErrNotfound {
```
Comment (zhuangqh) [inline cri/v1alpha2/cri.go]: unnecessary in CRI part. 1. useless HTTP type error here. 2. kubelet have not recognize the image `not found` error

File: cri/v1alpha2/cri_utils.go [hunk 5]
```diff
@@ -1006,6 +1008,9 @@ func (c *CriManager) ensureSandboxImageExists(ctx context.Context, imageRef stri
 	if errtypes.IsNotfound(err) {
 		err = c.ImageMgr.PullImage(ctx, imageRef, nil, bytes.NewBuffer([]byte{}))
 		if err != nil {
+			if err == errtypes.ErrNotfound {
```
Comment (zhuangqh) [inline cri/v1alpha2/cri_utils.go]: useless

File: ctrd/utils.go [hunk 6]
```diff
@@ -166,7 +165,8 @@ func (c *Client) getResolver(ctx context.Context, authConfig *types.AuthConfig,
 	}
 
 	if availableRef == "" {
-		return nil, "", fmt.Errorf("there is no available image reference after trying %+q", refs)
+		logrus.Errorf("there is no available image reference after trying %+q", refs)
+		return nil, "", errtypes.ErrNotfound
```
Comment (zhuangqh) [inline ctrd/utils.go]: prefer `errors.Wrapf(errtypes.ErrNotfound, "there is no available image reference after trying %+q", refs)` here
util function just directly return error, log or not should be decided by upper.

**Comment by zhuangqh** [comment]
could you show the UX when pulling a not-exist image? What is the error message display in terminal after doing `pouch pull xxx`. I am afraid some internal test would fail

**Comment by ZYecho** [comment]
> could you show me the UX when pulling a not-exist image? What is the error message display in terminal after doing `pouch pull xxx`. I am afraid some internal test would fail

```
sudo pouch pull fsadfsdafsdaf:fsdfdsaf
Error: failed to pull image: {"message":"not found"}


```

File: cri/v1alpha2/cri_utils.go (hunk 5 above)
Comment (ZYecho) [reply cri/v1alpha2/cri_utils.go]: ok

File: ctrd/utils.go (hunk 6 above)
Comment (ZYecho) [reply ctrd/utils.go]: I change to `logrus.warnf`, add an another log in upperand. But I think `errtypes.ErrNotfound` is enough here, user didn't care about how you find a available ref, just tell them not found.

Comment (zhuangqh) [reply ctrd/utils.go]: make sense

**Comment by ZYecho** [comment]
failed test try to pull `reg.docker.alibaba-inc.com/base/hello-world`， I try to pull it use docker， also didn't download it， seems it was an inner registry？

**Pull request merged**